db/util.go

```diff
@@ -0,0 +1,116 @@
+package db
+
+import (
+	"bytes"
+	"os"
+)
+
+func cp(bz []byte) (ret []byte) {
+	ret = make([]byte, len(bz))
+	copy(ret, bz)
+	return ret
+}
+
+// Returns a slice of the same length (big endian)
+// except incremented by one.
+// Returns nil on overflow (e.g. if bz bytes are all 0xFF)
+// CONTRACT: len(bz) > 0
+func cpIncr(bz []byte) (ret []byte) {
+	if len(bz) == 0 {
+		panic("cpIncr expects non-zero bz length")
+	}
+	ret = cp(bz)
+	for i := len(bz) - 1; i >= 0; i-- {
+		if ret[i] < byte(0xFF) {
+			ret[i]++
+			return
+		}
+		ret[i] = byte(0x00)
+		if i == 0 {
+			// Overflow
+			return nil
+		}
+	}
+	return nil
+}
+
+// IsKeyInDomain returns true iff a key lies within a domain.
+// See DB interface documentation for more information.
+func IsKeyInDomain(key, start, end []byte) bool {
+	if bytes.Compare(key, start) < 0 {
+		return false
+	}
+	if end != nil && bytes.Compare(end, key) <= 0 {
+		return false
+	}
+	return true
+}
+
+// FileExists returns true iff a file path exists.
+func FileExists(filePath string) bool {
+	_, err := os.Stat(filePath)
+	return !os.IsNotExist(err)
+}
+
+// Encapsulates the current valid versions and computes the next version.
+type VersionManager struct {
+	versions []uint64
+}
+
+func NewVersionManager(versions []uint64) *VersionManager {
+	return &VersionManager{versions: versions}
+}
+
+func (vm *VersionManager) Valid(version uint64) bool {
+	// todo: maybe use map to avoid linear search
```

Comment:
yeah, or use another binary search if the versions are sorted

Reply:
We can, but I would rather optimize this later

Reply:
ok, can you create an issue to track this work?

Reply:
It's a pretty simple change, so I will just update it to a map

Reply:
Exactly, it's very easy optimization, no need to "double" the work.